### PR TITLE
[TTI][RISCV]Improve costs for whole vector reg extract/insert.

### PR DESCRIPTION
**Pull request by alexey-bataev**
If we can detect, that whole register extract/insert is requested,
consider it free.

### CODE REVIEW

**Comment by alexey-bataev** [comment]
> Don't we need to know the exact vlen to know where register boundaries are?

I use getRegUsageForType() to get this info.

**Comment by topperc** [comment]
> > Don't we need to know the exact vlen to know where register boundaries are?
> 
> I use getRegUsageForType() to get this info.

That tells the maximum number of registers needed for the type assuming a minimum VLEN. If hardware VLEN is more than the minimum VLEN, we still use the extra registers but the elements in them are not used since they would be past VL. CodeGen has to use a slidedown unless we also know the maximum VLEN is the same as the minimum VLEN.

**Comment by alexey-bataev** [comment]
> > > Don't we need to know the exact vlen to know where register boundaries are?
> > 
> > 
> > I use getRegUsageForType() to get this info.
> 
> That tells the maximum number of registers needed for the type assuming a minimum VLEN. If hardware VLEN is more than the minimum VLEN, we still use the extra registers but the elements in them are not used since they would be past VL. CodeGen has to use a slidedown unless we also know the maximum VLEN is the same as the minimum VLEN.

Do we have anything in TTI that returns correct VLEN?

**Comment by topperc** [comment]
> > > > Don't we need to know the exact vlen to know where register boundaries are?
> > > 
> > > 
> > > I use getRegUsageForType() to get this info.
> > 
> > 
> > That tells the maximum number of registers needed for the type assuming a minimum VLEN. If hardware VLEN is more than the minimum VLEN, we still use the extra registers but the elements in them are not used since they would be past VL. CodeGen has to use a slidedown unless we also know the maximum VLEN is the same as the minimum VLEN.
> 
> Do we have anything in TTI that returns correct VLEN?

You can check that `ST->getRealMaxVlen() == ST->getRealMinVLen()`

**Comment by alexey-bataev** [comment]
> > > > > Don't we need to know the exact vlen to know where register boundaries are?
> > > > 
> > > > 
> > > > I use getRegUsageForType() to get this info.
> > > 
> > > 
> > > That tells the maximum number of registers needed for the type assuming a minimum VLEN. If hardware VLEN is more than the minimum VLEN, we still use the extra registers but the elements in them are not used since they would be past VL. CodeGen has to use a slidedown unless we also know the maximum VLEN is the same as the minimum VLEN.
> > 
> > 
> > Do we have anything in TTI that returns correct VLEN?
> 
> You can check that `ST->getRealMaxVlen() == ST->getRealMinVLen()`

Will add

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -326,6 +326,50 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
     default:
       break;
+    case TTI::SK_ExtractSubvector:
+      if (isa<FixedVectorType>(SubTp)) {
+        unsigned TpRegs = getRegUsageForType(Tp);
+        unsigned NumElems =
+            divideCeil(Tp->getElementCount().getFixedValue(), TpRegs);
+        // Whole vector extract - just the vector itself + (possible) vsetvli.
+        // TODO: consider adding the cost for vsetvli.
+        if (Index == 0 || (ST->getRealMaxVLen() == ST->getRealMinVLen() &&
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:336]: I think this check would be more clearly expressed as an and of the following clauses
a) ST->getRealMaxVLen() == ST->getRealMinVLen()
b) NumElems * ElementSizeInBits == VLEN
c) Index % NumElems == 0

Note that this only supports m1 full extracts.  But starting there and extending it to m2, and m4 later seems entirely reasonable.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -326,6 +326,50 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+                           Index % NumElems == 0)) {
+          std::pair<InstructionCost, MVT> SubLT =
+              getTypeLegalizationCost(SubTp);
+          return Index == 0
+                     ? TTI::TCC_Free
+                     : SubLT.first * getRISCVInstructionCost(RISCV::VMV_V_V,
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:342]: For a full VREG case, you never need the VMV_V_V.  You only need the VMV_V_V if NumElems < VLMAX.  

Extending this to sub-register extract with exact VLEN known would be reasonable, but let's do that in a separate patch.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -326,6 +326,50 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+                                                             SubLT.second,
+                                                             CostKind);
+        }
+      }
+      break;
+    case TTI::SK_InsertSubvector:
+      if (auto *FSubTy = dyn_cast<FixedVectorType>(SubTp)) {
+        unsigned TpRegs = getRegUsageForType(Tp);
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:350]: Same basic style comments as above.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 4]
```diff
@@ -326,6 +326,48 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+          LT.second.getVectorElementType() != MVT::i1) {
+        unsigned TpRegs = getRegUsageForType(Tp);
+        unsigned NumElems =
+            divideCeil(Tp->getElementCount().getFixedValue(), TpRegs);
+        // Whole vector extract - just the vector itself + (possible) vsetvli.
+        // TODO: consider adding the cost for vsetvli.
+        if (Index == 0 || (ST->getRealMaxVLen() == ST->getRealMinVLen() &&
+                           NumElems * LT.second.getScalarSizeInBits() ==
+                               ST->getRealMinVLen() &&
+                           Index % NumElems == 0))
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:340]: Should we add some test cases that exercise the exact VLEN path? I presume we'll need to add another RUN line with `-riscv-v-vector-bits-min=... -riscv-v-vector-bits-max=...`

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 5]
```diff
@@ -326,6 +326,48 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+    case TTI::SK_ExtractSubvector:
+      if (isa<FixedVectorType>(SubTp) &&
+          LT.second.getVectorElementType() != MVT::i1) {
+        unsigned TpRegs = getRegUsageForType(Tp);
+        unsigned NumElems =
+            divideCeil(Tp->getElementCount().getFixedValue(), TpRegs);
+        // Whole vector extract - just the vector itself + (possible) vsetvli.
+        // TODO: consider adding the cost for vsetvli.
+        if (Index == 0 || (ST->getRealMaxVLen() == ST->getRealMinVLen() &&
+                           NumElems * LT.second.getScalarSizeInBits() ==
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:338]: Should this be checking that the extracted subvector type is VLEN sized, not the containing vector type?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 4 above)
Comment (alexey-bataev) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Added couple, hope it works

File: llvm/test/Analysis/CostModel/RISCV/shuffle-extract_subvector.ll [hunk 6]
```diff
@@ -60,30 +61,55 @@ define void @test_vXi64(<4 x i64> %src256, <8 x i64> %src512) {
...
-; CHECK-NEXT:  Cost Model: Found an estimated cost of 4 for instruction: %V512_01234567 = shufflevector <16 x i32> %src512, <16 x i32> undef, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
-; CHECK-NEXT:  Cost Model: Found an estimated cost of 4 for instruction: %V512_89ABCDEF = shufflevector <16 x i32> %src512, <16 x i32> undef, <8 x i32> <i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15>
-; CHECK-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: ret void
+; RTBASE-LABEL: 'test_vXi32'
+; RTBASE-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: %V128_01 = shufflevector <4 x i32> %src128, <4 x i32> undef, <2 x i32> <i32 0, i32 1>
+; RTBASE-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %V128_23 = shufflevector <4 x i32> %src128, <4 x i32> undef, <2 x i32> <i32 2, i32 3>
+; RTBASE-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: %V256_01 = shufflevector <8 x i32> %src256, <8 x i32> undef, <2 x i32> <i32 0, i32 1>
+; RTBASE-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %V256_23 = shufflevector <8 x i32> %src256, <8 x i32> undef, <2 x i32> <i32 2, i32 3>
```
Comment (lukel97) [inline llvm/test/Analysis/CostModel/RISCV/shuffle-extract_subvector.ll:68]: Just making a note, this is an LMUL 1 vslidedown:

```asm
	vsetivli	zero, 2, e32, m1, ta, ma
	vslidedown.vi	v8, v8, 2
```

So I think the cost should just be one here. This looks like it's coming from the scalable vector cost path. Is the type being passed in a <8 x i32> instead of a <2 x i32>? Something that could be looked at in a later patch.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 7]
```diff
@@ -326,6 +326,48 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
     default:
       break;
+    case TTI::SK_ExtractSubvector:
+      if (isa<FixedVectorType>(SubTp) &&
+          LT.second.getVectorElementType() != MVT::i1) {
+        unsigned SubTpRegs = getRegUsageForType(SubTp);
+        unsigned SubNumElems = NextPowerOf2(
+            divideCeil(SubTp->getElementCount().getFixedValue(), SubTpRegs));
+        // Whole vector extract - just the vector itself + (possible) vsetvli.
+        // TODO: consider adding the cost for vsetvli.
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:336]: Generally we've just been ignoring vsetvli costs throughout TTI, so it's probably ok to leave out the todo

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 8]
```diff
@@ -326,6 +326,48 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+        unsigned NumElems = NextPowerOf2(
+            divideCeil(Tp->getElementCount().getFixedValue(), TpRegs));
+        // Whole vector insert - just the vector itself + (possible) vsetvli.
+        // TODO: consider adding the cost for vsetvli.
+        if ((Index == 0 || (ST->getRealMaxVLen() == ST->getRealMinVLen() &&
+                            NumElems * LT.second.getScalarSizeInBits() ==
+                                ST->getRealMinVLen() &&
+                            Index % NumElems == 0)) &&
+            (any_of(Args, UndefValue::classof) ||
+             (SubTpRegs != 0 && SubTpRegs != NextSubTpRegs &&
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:359]: When would SubTpRegs == 0? Is it possible for getRegUsageForType to return 0?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 9]
```diff
@@ -326,6 +326,48 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+        unsigned SubTpRegs = getRegUsageForType(SubTp);
+        unsigned SubNumElems = NextPowerOf2(
+            divideCeil(SubTp->getElementCount().getFixedValue(), SubTpRegs));
+        // Whole vector extract - just the vector itself + (possible) vsetvli.
+        // TODO: consider adding the cost for vsetvli.
+        if (Index == 0 || (ST->getRealMaxVLen() == ST->getRealMinVLen() &&
+                           SubNumElems * LT.second.getScalarSizeInBits() ==
+                               ST->getRealMinVLen() &&
+                           Index % SubNumElems == 0))
+          return TTI::TCC_Free;
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:341]: Style request - Please use early return for the Index == 0 case.

Once you do that, you can sink all of the other code inside a check for the exact VLEN and it should be much more clearly isolated.

**Review by preames — COMMENTED** [review]
I'd encourage you to split off a change to handle only the Index == 0 case.  It should be simple, but that's the value.  :)

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 8 above)
Comment (alexey-bataev) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:359]: Just to be safe here. I saw sometimes that these functions may return 0 in some corner cases (mostly for generic target), just kind of overprotection

**Comment by preames** [comment]
> I'd encourage you to split off a change to handle only the Index == 0 case. It should be simple, but that's the value. :)

I posted this here: https://github.com/llvm/llvm-project/pull/81751

**Comment by alexey-bataev** [comment]
The latest version of the patch actually handles only index 0

**Comment by lukel97** [comment]
#81751 handles both fixed and scalable vectors from the looks of things. I wonder if it's possible to have this patch handle the whole reg extract/insert case for scalable vectors too? Any scalable vector extract/insert should be free if both the vector and subvector are >= LMUL 1.

**Comment by lukel97** [comment]
> LMUL >=1 not enough for the second case, also need to check that the whole vector is insert, not, say, half of it.

But for llvm.vector.insert there is the constraint that all the subvec elements must be within bounds of the vector, and for scalable vectors the index at which it is inserted is scaled by vscale.

So if the subvector is LMUL >=1 it shouldn't be possible for only half of it be inserted, since it won't be truncated and the index will be a multiple of an LMUL1 register boundary.

This is separate from the exact VLEN fixed vector case in the original version of this PR though, we can leave it for a future patch.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 10]
```diff
@@ -326,6 +326,18 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
     default:
       break;
+    case TTI::SK_InsertSubvector:
+      if (auto *FSubTy = dyn_cast<FixedVectorType>(SubTp)) {
+        unsigned TpRegs = getRegUsageForType(Tp);
+        unsigned SubTpRegs = getRegUsageForType(SubTp);
+        unsigned NextSubTpRegs = getRegUsageForType(FixedVectorType::get(
+            Tp->getElementType(), FSubTy->getNumElements() + 1));
+        // Whole vector insert - just the vector itself.
+        if (Index == 0 && SubTpRegs != 0 && SubTpRegs != NextSubTpRegs &&
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:336]: Just want to check, is `SubTpRegs != NextSubTpRegs` to check that SubTp isn't a fractional LMUL?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 11]
```diff
@@ -326,6 +326,18 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+            TpRegs / SubTpRegs > 1)
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:337]: Should this be
```suggestion
            TpRegs / SubTpRegs >= 1)
```

So inserts with the same vector and subvector type are also free? E.g @llvm.vector.insert.v128i8.v128i8 in the tests is still 8

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 12]
```diff
@@ -432,12 +444,19 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+    if (Index == 0)
+      return TTI::TCC_Free;
+
     // Example sequence:
     // vsetivli     zero, 4, e8, mf2, tu, ma (ignored)
     // vslidedown.vi  v8, v9, 2
     return LT.first *
            getRISCVInstructionCost(RISCV::VSLIDEDOWN_VI, LT.second, CostKind);
   case TTI::SK_InsertSubvector:
+    if (Index == 0 && any_of(Args, UndefValue::classof))
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:457]: Now that you mention it in https://github.com/llvm/llvm-project/pull/81818#discussion_r1492834518, the args here aren't necessarily the vector and subvector right? I think they're the arguments to the shufflevector instruction.

So maybe there isn't an easy way to handle the scalable insert case after all.

Comment (alexey-bataev) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:457]: If we know that it is a subvector insert and at least one argument is undef vector, does it really matter, which one is undef? Iа the subvector is undef - we can use the original vector, in other case - just resize the subvector to the type of the vector.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 11 above)
Comment (alexey-bataev) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:337]: Ok, will do

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 12 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:457]: Yeah, I think that's what RISCVISelLowering currently does.

My concern was that it Args could be arbitrary inserts from possibly the same vector, e.g.

```llvm
shufflevector <8 x i32> %x, <8 x i32> poison, <4 x i32> <i32 0, i32, 1, i32 4, i32 5>
```
Would be an insert of the second half of %x into the first half of %x.

But I took a closer look and I don't think TTI::SK_InsertSubvector is actually used in this case. `ShuffleVectorInst::isInsertSubvectorMask` only seems to detect inserts of operand 2 into operand 1, so I think the args can be treated as vector and subvector respectively.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 13]
```diff
@@ -442,6 +454,9 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:457]: Do we need to check that Args isn't empty?

Comment (alexey-bataev) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Added, thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 14]
```diff
@@ -326,6 +326,18 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
     switch (Kind) {
     default:
       break;
+    case TTI::SK_InsertSubvector:
+      if (auto *FSubTy = dyn_cast<FixedVectorType>(SubTp)) {
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:330]: Sorry just noticing this now, I think this should be a cast since it shouldn't be possible to insert a scalable `SubTp` into `Tp`
```suggestion
      auto *FSubTy = cast<FixedVectorType>(SubTp);
```

Comment (alexey-bataev) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:330]: Ok, will co, tried to add an extra level of protection here

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 15]
```diff
@@ -326,6 +326,18 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
     switch (Kind) {
     default:
       break;
+    case TTI::SK_InsertSubvector: {
+      auto *FSubTy = dyn_cast<FixedVectorType>(SubTp);
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:330]: Should we use cast instead of dyn_cast here to get the assertion?

**Review by lukel97 — COMMENTED** [review]
Does the PR title need updated to reflect that this handles insert_subvectors only now

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 16]
```diff
@@ -442,6 +454,9 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
     return LT.first *
            getRISCVInstructionCost(RISCV::VSLIDEDOWN_VI, LT.second, CostKind);
   case TTI::SK_InsertSubvector:
+    if (Index == 0 && !Args.empty() && any_of(Args, UndefValue::classof))
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:457]: I tried to split off this piece - or more accurately something vaguely related - and stumbled into something interesting.

The InsertSubvector w/Index=0 is unreachable from everywhere except SLP.  TTI::getInstructionCost contains a check for the identity shuffle and always returns 0.  improveShuffleKindFromMask will recognize the insert into passthru case as a select (correctly), and thus it doesn't hit this case either.  Put together, this means that the index=0 case never makes it from the backend, and thus we have no test coverage via cost model tests.

SLP hits a slightly different codepath here and directly calls getShuffleCost with a possible identity mask.  It still can't hit the select case, but it can hit the insert into poison case.  SLP appears to have a bunch of guards for this already in various cases.

I'm not really a fan of having untestable logic here.  Anyone have any ideas how we can rework this API to ensure SLP can't reach a case which is untestable via costmodel tests?

Comment (alexey-bataev) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:457]: Do not tests for llvm.vector.insert intrinsics check this?

**Comment by alexey-bataev** [comment]
Ping!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 17]
```diff
@@ -326,6 +326,18 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
     default:
       break;
+    case TTI::SK_InsertSubvector: {
+      auto *FSubTy = cast<FixedVectorType>(SubTp);
+      unsigned TpRegs = getRegUsageForType(Tp);
+      unsigned SubTpRegs = getRegUsageForType(SubTp);
+      unsigned NextSubTpRegs = getRegUsageForType(FixedVectorType::get(
+          Tp->getElementType(), FSubTy->getNumElements() + 1));
+      // Whole vector insert - just the vector itself.
+      if (Index == 0 && SubTpRegs != 0 && SubTpRegs != NextSubTpRegs &&
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:336]: I don't think this works. getRegUsageForType is returning the maximum number of registers needed given a minimum VLEN. If the runtime VLEN is larger the used number of registers could be less.

The backend must always use a vslideup for fixed vector insert unless we know both the maximum and minimum VLEN are the same. I think you have to check `ST.getRealVLen()`.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 18]
```diff
@@ -457,12 +457,28 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+    const unsigned MinVLen = ST->getRealMinVLen();
+    const unsigned MaxVLen = ST->getRealMaxVLen();
+    // Whole vector insert - just the vector itself.
+    if (auto *FSubTy = dyn_cast<FixedVectorType>(SubTp);
+        FSubTy && Index == 0 && MinVLen == MaxVLen) {
+      unsigned TpRegs = getRegUsageForType(Tp);
+      unsigned SubTpRegs = getRegUsageForType(SubTp);
+      unsigned NextSubTpRegs = getRegUsageForType(FixedVectorType::get(
+          Tp->getElementType(), FSubTy->getNumElements() + 1));
+      if (SubTpRegs != 0 && SubTpRegs != NextSubTpRegs && TpRegs >= SubTpRegs)
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:472]: Is it possible for `TpRegs < SubTpRegs`?

**Review by lukel97 — COMMENTED** [review]
I think the PR title still needs the reg extract part removed

**Review by preames — CHANGES_REQUESTED** [review]
I have posted an alternative patch for the remaining insertsubvector case here: https://github.com/llvm/llvm-project/pull/85240

When writing this, I discovered that this patch models a lowering which is not implemented.  There is a patch on review, but it hasn't landed yet.
